Slack on change only

ctro opened the pull request:
OK, I did a thing.

I wanted to see how much more readable the summary code would be if I class-ified it.  Turns out it's way more readable :)  Need to stop watching Rich Hickey talks.

Then I made Slack notify on change only, which was as simple as NOT writing a summary.txt file if there was no change.  This is not (yet?) configurable at all.

THEN I ran into the "NO CHANGE" bug and realized it was related to OpenOpps not having any zap issues.  So I changed the way that is handled.  I'm still not sure if OpenOpps _really does_ have 0 zap issues, or if zap is zapping out.

Tons of RED/GREEN here :(  But not really huge changes.

Relates to 
https://trello.com/c/UnfzMVR9/111-a-user-should-be-able-to-configure-slack-notification-presence-frequency-and-options
and
https://trello.com/c/y63dRvhw/133-investigate-c2-scans-always-showing-as-new-results

Posting for feedback now, I can get back to this next week.
Also, could merge Aidan's S3 thing first and I could deal with whatever merge needs to be done here.

ctro committed:
Add require for Tempfile

ctro committed:
Temp local targets

ctro committed:
Only notify on change

ctro commented on tasks/summarize-zap-results/task.rb:
```diff
@@ -11,8 +11,12 @@ def project_name
 puts "Comparing last results to current results..."
-ZAPResultSetComparator.write_summary(last_run_dir, current_run_dir, output_dir, project_name)
+comparator = ZAPResultSetComparator.new(project_name, last_run_dir, current_run_dir)
+if comparator.no_change?
```
Here's where we skip writing the `summary.txt` if there is no change

afeld replied:
Does that prevent it from posting in Slack, or will it just be an empty message?

ctro replied:
It prevents it from posting in slack:

> when text is omitted, and...  text_file omitted, or present but file missing: > (no notification given)

From https://github.com/cloudfoundry-community/slack-notification-resource

I guess the behavior of that resource could change on us.

afeld replied:
Nope, works for me! Maybe just link to https://github.com/cloudfoundry-community/slack-notification-resource#parameters and mention it in an inline comment?

ctro commented on tasks/summarize-zap-results/test/test_zap_result_set.rb:
```diff
@@ -7,42 +7,50 @@ class TestZAPResultSet < MiniTest::Test
+    describe ".missing?" do
+      it "should return true if the project file is missing" do
+        refute ZAPResultSet.new('fake-site-1', curr_results_dir).missing?
+        refute ZAPResultSet.new('fake-site-0', curr_results_dir).missing?
```
Here's the test for the OpenOpps "0 zap issues" reporting NEW SITE

afeld replied:
Maybe worth having an "integration" test for it? i.e. that the message should say "no change" instead of "new site".

ctro commented:
If this is _too much_ I can split the commits out.  Mostly wanted to get this out there before the weekend started.

ctro committed:
codeclimate fixes

afeld commented:
> I wanted to see how much more readable the summary code would be if I class-ified it. Turns out it's way more readable :) Need to stop watching Rich Hickey talks.

:laughing:

> Then I made Slack notify on change only...This is not (yet?) configurable at all.

That seems like a reasonable default.

> I'm still not sure if OpenOpps really does have 0 zap issues

Maybe try running the scan locally through desktop ZAP?

> If this is too much I can split the commits out.

If it's easy to do so, that might help keep the PR comments focused, but not a huge deal otherwise.

afeld commented on pipelines/zap/pipeline.yml:
```diff
@@ -88,9 +88,9 @@ jobs:
       put: slack
       params:
         channel: <%= project['slack_channel'] ? "'##{project['slack_channel']}'" : "{{slack-channel}}" %>
-        icon_url: {{slack-icon}}
```
Mind removing this from the `config/*.example.yml` files, and below?

ctro replied:
I didn't really mean to push this part. It does work though. Is an emoji better than an icon_url?  If so I'll just adjust the {{config}}

afeld replied:
No strong feelings. We probably shouldn't be using CloudApp to host the image for the production pipeline anyway.

https://github.com/18F/concourse-compliance-testing/blob/9bd27272a014eaa4ee08a0810fe43074c97c5ce4/config/prod.example.yml#L13

afeld replied:
I would just take out the configuration altogether. Convention over configuration 😉

ctro committed:
Address Feedback
- slack-icon-emoji
- rubocop formatting
- document s3 resource usage
- integration test for NO CHANGE

ctro committed:
fix indentation
